File: pruebas/JacobiMethod/jm.c
#define _POSIX_C_SOURCE 199309L
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>
#include <time.h>
#include <stdint.h>
#include <unistd.h>

void load_data(double *a, double *b, int n) {
    FILE *fa = fopen("a_matrix.bin", "rb");
    if (!fa) {
        perror("Error opening a_matrix.bin");
        exit(EXIT_FAILURE);
    }

    int r1= fread(a, sizeof(double), n * n, fa);
    fclose(fa);

    FILE *fb = fopen("b_vector.bin", "rb");
    if (!fb) {
        perror("Error opening b_vector.bin");
        exit(EXIT_FAILURE);
    }
    int r2=fread(b, sizeof(double), n, fb);
    fclose(fb);
}

void jacobi(int n, int max_iter, double tol) {

    double total_time_sum = 0.0;

    MPI_Init(NULL, NULL);
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm comm = MPI_COMM_WORLD;

    if (n % size != 0) {
        n = ((n + size - 1) / size) * size;
    }

    double *a = aligned_alloc(64, n * n * sizeof(double));
    double *x = aligned_alloc(64, n * sizeof(double));
    double *b = aligned_alloc(64, n * sizeof(double));
    
    if (rank == 0) {
        load_data(a,b,n);
    }

    double *x_new = malloc(n * sizeof(double));
    double norm = 0.0;
    int iter = 0;

    struct timespec t_start, t_end;
    MPI_Barrier(MPI_COMM_WORLD);
    clock_gettime(CLOCK_MONOTONIC, &t_start);  

    MPI_Bcast(a, n * n, MPI_DOUBLE, 0, comm);
    MPI_Bcast(b, n, MPI_DOUBLE, 0, comm);

    int chunk_size = (int)ceil((double)n / (double)size);
    int start = rank * chunk_size;
    int end_idx = (start + chunk_size > n) ? n : start + chunk_size;

    while (iter < max_iter) {
        norm = 0.0;
        for (int i = start; i < end_idx; i++) {
            double sigma = 0.0;
            for (int j = 0; j < i; j++) {
                sigma += a[i * n + j] * x[j];
            }
            for (int j = i + 1; j < n; j++) {
                sigma += a[i * n + j] * x[j];
            }
            x_new[i] = (b[i] - sigma) / a[i * n + i];
            norm += (x_new[i] - x[i]) * (x_new[i] - x[i]);
        }

        MPI_Allreduce(MPI_IN_PLACE, &norm, 1, MPI_DOUBLE, MPI_SUM, comm);

        if (sqrt(norm) < tol) {
            break;
        }

        MPI_Allgather(x_new + start, end_idx - start, MPI_DOUBLE,x, end_idx - start, MPI_DOUBLE, comm);
        iter++;
    }

    clock_gettime(CLOCK_MONOTONIC, &t_end);
    double total_time=(t_end.tv_sec - t_start.tv_sec) + (t_end.tv_nsec - t_start.tv_nsec) / 1e9;
    MPI_Reduce(&total_time, &total_time_sum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

    if (rank == 0) {
        FILE *f = fopen("../Resultados/JM/JMC.txt", "a");
        if (f != NULL) {
            fprintf(f, "%d,%d,%.9f\n", size, n, total_time_sum / size);
            fclose(f);
        }
    }
    
    if (rank==0){
        free(a); 
        free(b); 
        free(x); 
        free(x_new);
    }

    MPI_Finalize();
}

int main(int argc, char** argv) {

    int n = atoi(argv[1]);
    int i = atoi(argv[2]);
    double t = atof(argv[3]);

    jacobi(n, i, t);
    return 0;
}

File: pruebas/JacobiMethod/jm.jl
include("../bindings/macroBindings.jl")
include("../bindings/functionBindings.jl")

using LinearAlgebra, Random, Libdl, Printf

function load_data!(a::Matrix{Float64}, b::Vector{Float64})
    open("a_matrix.bin", "r") do io
        read!(io, a) 
    end
    open("b_vector.bin", "r") do io
        read!(io, b)
    end
    return a, b
end

function jacobi(n::Int, max_iter::Int, tol::Float64)::Tuple{Float64, Int}

    total_time_sum = Ref{Cdouble}(0.0)
    p=-1
    t=0.0

    MPI_Init()
    rank = Ref{Cint}()
    size = Ref{Cint}()
    comm = MPI_COMM_WORLD
    double = MPI_DOUBLE
    MPI_Comm_rank(comm, rank)
    MPI_Comm_size(comm, size)

    if n % size[] != 0
        n = ceil(Int, n / size[]) * size[]
    end
    
    a = Matrix{Float64}(undef, n, n)
    x = zeros(Float64, n)
    b = Vector{Float64}(undef, n)

    if rank[]==0
        a,b=load_data!(a,b)
    end
    
    x_new = similar(x)         
    nor = Ref{Float64}(0.0)   
    iter = 0  

    MPI_Barrier(MPI_COMM_WORLD)
    t_start = time_ns()

    MPI_Bcast(a, Cint(n*n), double, Cint(0), comm)
    MPI_Bcast(b, Cint(n), double, Cint(0), comm)

    chunk_size = ceil(Int, n / size[])           
    start = rank[] * chunk_size + 1                
    end_idx = min(start + chunk_size - 1, n)   

    while iter < max_iter
        nor[] = 0.0
        for i in start:end_idx
            sigma = 0.0
            for j in 1:i-1
                sigma += a[j,i] * x[j]
            end
            for j in i+1:n
                sigma += a[j,i] * x[j]
            end
            x_new[i] = (b[i] - sigma) / a[i,i]   
            nor[] += (x_new[i] - x[i])*(x_new[i] - x[i])   
        end

        MPI_Allreduce(MPI_IN_PLACE, nor, Cint(1), double, MPI_SUM, comm)

        if sqrt(nor[]) < tol
            break
        end

        MPI_Allgather(pointer(x_new,start), Cint(end_idx - start + 1), double,x, Cint(end_idx - start + 1), double, comm)
        iter += 1
    end

    t_end = time_ns()
    total_time=(t_end - t_start)/1e9
    MPI_Reduce(Ref(total_time), total_time_sum, Cint(1), MPI_DOUBLE, MPI_SUM, Cint(0), MPI_COMM_WORLD)

    if rank[] == 0
        p=size[]
        t=total_time_sum[] / size[] 
    end
    
    MPI_Finalize()

    return (t,p)

end

n = parse(Int, ARGS[1])
r = parse(Int, ARGS[2])
tol = parse(Float64, ARGS[3])

t,p=jacobi(n, r, tol)

if p!=-1
    open("../Resultados/JM/JMMPI4ALL.txt", "a") do file
        write(file,string(p, ",", n, ",", t, "\n"))
    end
end



